src/main/go/org/jenkinsci/plugins/durabletask/heartbeat-launcher.go

```diff
@@ -0,0 +1,165 @@
+package main
+
+import (
+	"fmt"
+	"os"
+	"os/exec"
+	"os/signal"
+	"sync"
+
+	"golang.org/x/sys/unix"
+)
+
+// Note: EXIT signal is needed as some shell variants require an EXIT trap after catching a signal
+const trapSig = "trap ':' INT TERM EXIT"
+
+func checkHeartbeatErr(err error) {
+	checkErr("heartbeat", err)
+}
+
+func checkLauncherErr(err error) {
+	checkErr("launcher", err)
+}
+
+func checkErr(process string, err error) {
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "(%v) check err: %v\n", process, err.Error())
+	}
+}
+
+// Catch a termination signal to allow for a graceful exit (i.e. no zombies)
+func signalCatcher(sigChan chan os.Signal) {
+	for sig := range sigChan {
+		fmt.Printf("(sig catcher) caught: %v\n", sig)
+	}
+}
+
+func launcher(wg *sync.WaitGroup, pidChan chan int,
+	scriptString string, logPath string, resultPath string, outputPath string) {
+
+	defer wg.Done()
+	recordExit := fmt.Sprintf("status=\"$?\"; echo \"$status\" > %v.tmp; mv %v.tmp %v; wait; exit \"$status\"",
+		resultPath, resultPath, resultPath)
+	scriptWithExit := trapSig + "; " + scriptString + "; " + recordExit
+	scriptCmd := exec.Command("/bin/sh", "-c", scriptWithExit)
```

Comment:
Huh?? The whole point of using a Golang binary is to reimplement this functionality without shell scripts.

Reply:
So I originally had everything as goroutines, but then discovered that my main golang program was crashing whenever i terminated the Jenkins instance. This then prevented the ability to write out the exit status.  I think ended up converting everything to scripts to survive Jenkins instance terminations.

However, I did discover that a lot of my crashes were caused because I was piping to stdout, so when the Jenkins instance got terminated, it also crashed the binary.  I can most likely revisit this and convert back to all goroutines again.

Reply:
Yes, the hope was for the agent JVM to launch-and-forget a Golang binary, which in turn would directly launch the user process (typically `sh` but might be something else depending on `#!` and `Shell.DescriptorImpl.shell`) without any other helper processes and using only well-supported Golang system APIs.